synapse/app/synchrotron.py

```diff
@@ -236,6 +236,13 @@ def stream_positions(self):
         return {"typing": self._latest_room_serial}
 
     def process_replication_rows(self, token, rows):
+        if self._latest_room_serial > token:
+            # The master has gone backwards. To prevent inconsistent data, just
+            # clear everything.
+            self._room_serials = {}
+            self._room_typing = {}
```

Comment:
Would it be possible to move this out into a separate `_clear()` or `_reset()` function? Just so that its clearer that the typing handler *can* be reset